File: src/1349-Maximum-Students-Taking-Exam/1349.cpp
class Solution 
{
public:
    int maxStudents(vector<vector<char>>& seats) 
    {
        r = seats.size(), c = seats[0].size();
        match = vector<vector<int>>(r * c);
        int res = 0, cnt = 0;
        
        int vis[r * c] = {};
        for (int i = 0; i < r; i++)
        {
            for (int j = 0; j < c; j += 2)
            {
                if (seats[i][j] != '.') continue;
                memset(vis, 0, sizeof vis);
                vector<int> node = {i, j};
                if (find(node, vis, seats)) res++;
            }
        }
        
        for (int i = 0; i < r; i++)
            for (int j = 0; j < c; j++)
                if (seats[i][j] == '.') cnt++;
        return cnt - res;
    }
private:
    int d[6][2] = {{-1, -1}, {0, -1}, {1, -1}, {-1, 1}, {0, 1}, {1, 1}};
    int r, c;
    vector<vector<int>> match;
    
    bool find(vector<int>& node, int* vis, vector<vector<char>>& seats)
    {
        for (auto& it : d)
        {
            int nx = it[0] + node[0], ny = it[1] + node[1];
            if (nx >= 0 && nx < r && ny >= 0&& ny < c && !vis[nx * c + ny] && seats[nx][ny] == '.')
            {
                vis[nx * c + ny] = 1;
                if (match[nx * c + ny].empty() || find(match[nx * c + ny], vis, seats))
                {
                    match[nx * c + ny] = node;
                    return true;
                }
            }
        }
        return false;
    }
};

File: src/1349-Maximum-Students-Taking-Exam/1349.py
class Solution:
    def maxStudents(self, seats: List[List[str]]) -> int:
        r, c = len(seats), len(seats[0])
        match = [[-1] * c for _ in range(r)]
        
        def find(node, vis):
            x, y = node
            for i, j in [[-1, -1], [0, -1], [1, -1], [-1, 1], [0, 1], [1, 1]]:
                nx, ny = i + x, j + y
                if 0 <= nx < r and 0 <= ny < c and not vis[nx][ny] and seats[nx][ny] == '.':
                    vis[nx][ny] = True
                    if match[nx][ny] == -1 or find(match[nx][ny], vis):
                        match[nx][ny] = node
                        return True
            return False
        
        res, cnt = 0, 0
        for i in range(r):
            for j in range(0, c, 2):
                if seats[i][j] != '.': continue
                vis = [[False] * c for _ in range(r)]
                if find((i, j), vis): res += 1
        
        for i in range(r):
            for j in range(c):
                if seats[i][j] == '.': cnt += 1
        return cnt - res